polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -692,6 +906,138 @@ impl<T: Config> OnNewHead for Pallet<T> {
 	}
 }
 
+impl<T: Config> PreCodeUpgrade for Pallet<T> {
+	/// Ensures that all upgrade-related costs are covered for the specific parachain. Upon success,
+	/// it updates the deposit-related state for the parachain.
+	///
+	/// There are three cases where we do not charge any upgrade costs from the initiator
+	/// of the upgrade:
+	///
+	/// 1. `requirements` is explicitly set to `UpgradeRequirements::SkipRequirements`.
+	///
+	/// 2. System parachains do not pay for upgrade fees.
+	///
+	/// 3. All lease-holding parachains are permitted to do upgrades for free. This is introduced to
+	///    avoid causing a breaking change to the system once para upgrade fees are required.
+	fn pre_code_upgrade(
+		para: ParaId,
+		new_code: ValidationCode,
+		requirements: UpgradeRequirements,
+	) -> Result<Weight, Weight> {
+		let Some(head) = paras::Pallet::<T>::para_head(para) else {
+			return Err(T::DbWeight::get().reads(1))
+		};
+
+		let Some(mut info) = Paras::<T>::get(para) else { return Err(T::DbWeight::get().reads(2)) };
+		let lease_holding = Self::is_parachain(para);
+
+		// The billing account doesn't need to be set for lease holding parachains.
+		if info.billing_account.is_none() && !lease_holding {
+			Self::deposit_event(Event::<T>::CodeUpgradeScheduleFailed(
+				CodeUpgradeScheduleError::BillingAccountNotSet,
+			));
+			return Err(T::DbWeight::get().reads(3))
+		}
+
+		let per_byte_fee = T::DataDepositPerByte::get();
+		let new_deposit = T::ParaDeposit::get()
+			.saturating_add(per_byte_fee.saturating_mul((head.0.len() as u32).into()))
+			.saturating_add(per_byte_fee.saturating_mul((new_code.0.len() as u32).into()));
+
+		let current_deposit = info.deposit;
+
+		let free_upgrade = requirements == UpgradeRequirements::SkipRequirements ||
+			para.is_system() ||
+			lease_holding;
+
+		match (free_upgrade, info.billing_account.clone()) {
+			(false, Some(billing_account)) => {
+				if let Err(_) = <T as Config>::Currency::withdraw(
```

Comment:
Nit: Would be nice to have the business logic less deeply nested.

Reply:
Should be better now :)